### PR TITLE
feat: optimize enumerable mappings

### PR DESCRIPTION
**Pull request by TimothyMakkison**
# Replace linq usage with `for` and `foreach` loops

## Description
- Updated logic for implicit enumerable conversions, used when mapping synthetic.
- For known length sources use a `foreach` loop to construct an array or list. 
   - `Array`, `IReadOnlyList`, `IReadOnlyCollection`, `IEnumerable` -> `Array`
   - `List`, `IList`, `ICollection` -> `List`
- Keep linq logic for projection mapping and for `IEnumerable` source/target mapping.

Related #433

## Checklist

- [ ] The existing code style is followed
- [ ] The commit message follows our guidelines
- [ ] Performed a self-review of my code
- [ ] Hard-to-understand areas of my code are commented
- [ ] Unit tests are added/updated
- [ ] Integration tests are added/updated (as applicable, especially if feature/bug depends on roslyn or framework version in use)

### CODE REVIEW

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs [hunk 1]
```diff
@@ -127,6 +112,104 @@ ForEachAddEnumerableExistingTargetMapping CreateForEach(string methodName)
...
+
+        return null;
+    }
+
+    private static TypeMapping? TryBuildFastConversion(MappingBuilderContext ctx, ITypeMapping elementMapping)
+    {
+        if (ctx.IsExpression || !ctx.CollectionInfos!.Source.CountIsKnown)
+            return null;
+
+        // if array is same as or convertible to target
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:158]: i don't understand this comment 🤔

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs [hunk 2]
```diff
@@ -127,6 +112,104 @@ ForEachAddEnumerableExistingTargetMapping CreateForEach(string methodName)
...
+        {
+            {
+                return new CastMapping(ctx.Source, ctx.Target);
+            }
+        }
+
+        return null;
+    }
+
+    private static TypeMapping? TryBuildFastConversion(MappingBuilderContext ctx, ITypeMapping elementMapping)
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:153]: IMO this method gets a lot easier to read with a few little restructurings:
* move the `if (ctx.CollectionInfos.Target.CollectionType is CollectionType.List or CollectionType.ICollection or CollectionType.IList)` part up (just after the early null return)
* inverse the `ctx.CollectionInfos.Target.CollectionType` check and return null early
* inverse the `ctx.CollectionInfos.Source.CollectionType == CollectionType.Array` condition and return the `ArrayForEachMapping` early

File: src/Riok.Mapperly/Descriptors/Mappings/ArrayForEachMapping.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+using static Riok.Mapperly.Emit.SyntaxFactoryHelper;
+
+namespace Riok.Mapperly.Descriptors.Mappings;
+
+public class ArrayForEachMapping : MethodMapping
+{
+    private const string TargetVariableName = "target";
+    private const string LoopItemVariableName = "item";
+    private const string LoopCounterName = "i";
+    private const string ArrayLengthProperty = nameof(Array.Length);
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Mappings/ArrayForEachMapping.cs:14]: This is unused, isn't it?

File: src/Riok.Mapperly/Descriptors/Mappings/ArrayForEachMapping.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+using static Riok.Mapperly.Emit.SyntaxFactoryHelper;
+
+namespace Riok.Mapperly.Descriptors.Mappings;
+
+public class ArrayForEachMapping : MethodMapping
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Mappings/ArrayForEachMapping.cs:9]: Add a comment explaining this mapping...

File: src/Riok.Mapperly/Descriptors/Mappings/ArrayForEachMapping.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+        // target[i] = Map(item);
+        var (loopItemCtx, loopItemVariableName) = ctx.WithNewSource(LoopItemVariableName);
+        var convertedSourceItemExpression = _elementMapping.Build(loopItemCtx);
+
+        var assignment = Assignment(
+            ElementAccess(IdentifierName(targetVariableName), IdentifierName(loopCounterVariableName)),
+            convertedSourceItemExpression
+        );
+        // i++;
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Mappings/ArrayForEachMapping.cs:59]: Add a newline before a single line comment

File: src/Riok.Mapperly/Descriptors/Mappings/ForEachAddEnumerableMapping.cs [hunk 6]
```diff
@@ -14,24 +14,36 @@ namespace Riok.Mapperly.Descriptors.Mappings;
...
     public ForEachAddEnumerableMapping(
         ITypeSymbol sourceType,
         ITypeSymbol targetType,
         ITypeMapping elementMapping,
         ObjectFactory? objectFactory,
         string insertMethodName,
-        EnsureCapacity? ensureCapacityBuilder
+        EnsureCapacity? ensureCapacityBuilder,
+        ITypeSymbol? typeToInstantiate = null,
+        string? countPropertyName = null
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Mappings/ForEachAddEnumerableMapping.cs:28]: Add a comment explaining that this property can only be set if `typeToInstantiate` has a constructor with exact one int parameter. If this parameter is set, the ensure capacity isn't needed, is it?

File: test/Riok.Mapperly.Tests/Mapping/EnumerableDeepCloningTest.cs [hunk 7]
```diff
@@ -206,7 +206,22 @@ public void EnumerableOfPrimitivesToEnumerableDeepCloning()
...
+            var target = new int[source.Count];
+            var i = 0;
+            foreach (var item in source)
+            {
+                target[i] = item;
+                i++;
+            }
+
+            return target;
+            """
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/Mapping/EnumerableDeepCloningTest.cs:223]: Weird indention... (the same for a lot of the other inlined assertions)

File: test/Riok.Mapperly.Tests/Mapping/EnumerableDeepCloningTest.cs [hunk 8]
```diff
@@ -206,7 +206,22 @@ public void EnumerableOfPrimitivesToEnumerableDeepCloning()
     public void ReadOnlyCollectionOfPrimitivesToEnumerableDeepCloning()
     {
         var source = TestSourceBuilder.Mapping("IReadOnlyCollection<int>", "IEnumerable<int>", TestSourceBuilderOptions.WithDeepCloning);
-        TestHelper.GenerateMapper(source).Should().HaveSingleMethodBody("return global::System.Linq.Enumerable.ToArray(source);");
+        TestHelper
+            .GenerateMapper(source)
+            .Should()
+            .HaveSingleMethodBody(
+                """
+            var target = new int[source.Count];
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/Mapping/EnumerableDeepCloningTest.cs:214]: Wouldn't here a array clone be faster?

**Review by latonz — CHANGES_REQUESTED** [review]

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs [hunk 9]
```diff
@@ -127,6 +112,104 @@ ForEachAddEnumerableExistingTargetMapping CreateForEach(string methodName)
         }
     }
 
+    private static TypeMapping? TryBuildImplicitMapping(MappingBuilderContext ctx, ITypeMapping elementMapping)
```
[the remark this answers is not in the document]
Comment (TimothyMakkison) [reply src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:115]: Renamed. 

I can't use the simplified method as a guard must be used to prevent `int[]` -> `ICollection<int>` conversions

File: test/Riok.Mapperly.Tests/Mapping/EnumerableDeepCloningTest.cs (hunk 8 above)
Comment (TimothyMakkison) [reply test/Riok.Mapperly.Tests/Mapping/EnumerableDeepCloningTest.cs:214]: `IReadOnlyCollection` can't use `Array.Clone`.

It did make me think about `ICollection.CopyTo`... I imagine in some cases using `ToArray` will be faster for synthetic mappings if the source implements `ICollection`.

For instance deep cloning `IReadOnlyCollection<int>` -> `int[]` with `ToArray` is probably faster than a `for` loop as the source will likely be `ICollection` or array and will use `Array.Clone` or `ICollection.CopyTo` internally.

File: src/Riok.Mapperly/Descriptors/Mappings/ForEachAddEnumerableMapping.cs (hunk 6 above)
Comment (TimothyMakkison) [reply src/Riok.Mapperly/Descriptors/Mappings/ForEachAddEnumerableMapping.cs:28]: Should I create a separate constructor? `ForEachSetDictionaryMapping` probably needs one as well.

Comment (latonz) [reply src/Riok.Mapperly/Descriptors/Mappings/ForEachAddEnumerableMapping.cs:29]: A separate constructor may be a good solution.

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (latonz) [reply src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs]: Ahhhh I once more missed those runtime implementations for arrays... Could you mention sth. like this in the comment `Array implements IList and ICollection at runtime, see https://stackoverflow.com/q/47361775/3302887`.

**Comment by TimothyMakkison** [comment]
After benchmarking I decided to use `ToArray` and `ToList` for synthetic mappings. Chances are most Source values will implement `ICollection` and can use the fast `CopyTo` method. This is usually a copy of one span of memory into another, outperforming a `for` loop.

### tldr
Non synthetic/deep cloning -> `for`/`foreach` loops
Synthetic -> `ToArray`/`ToList`

### Benchmarks

#### `int` -> `int`
|                Method |  Size |      Mean |     Error |    StdDev | Allocated |
|---------------------- |------ |----------:|----------:|----------:|----------:|
| ArrayForEachAddToList | 10000 | 20.275 us | 0.1286 us | 0.1074 us |  39.12 KB |
|           ArrayToList | 10000 |  3.677 us | 0.0223 us | 0.0186 us |  39.12 KB |
|        ListForToArray | 10000 | 10.867 us | 0.0799 us | 0.0708 us |  39.09 KB |
|    ListSpanForToArray | 10000 |  9.126 us | 0.1420 us | 0.1186 us |  39.09 KB |
|           ListToArray | 10000 |  3.578 us | 0.0257 us | 0.0227 us |  39.09 KB |

#### `string` -> `string`
|                Method |  Size |      Mean |     Error |    StdDev | Allocated |
|---------------------- |------ |----------:|----------:|----------:|----------:|
| ArrayForEachAddToList | 10000 | 44.476 us | 0.8462 us | 0.7915 us |  78.18 KB |
|           ArrayToList | 10000 | 10.362 us | 0.1835 us | 0.1716 us |  78.18 KB |
|        ListForToArray | 10000 | 21.607 us | 0.2884 us | 0.2557 us |  78.15 KB |
|    ListSpanForToArray | 10000 | 38.190 us | 0.2374 us | 0.2220 us |  78.15 KB |
|           ListToArray | 10000 |  7.782 us | 0.0404 us | 0.0337 us |  78.15 KB |

File: src/Riok.Mapperly/Descriptors/Mappings/ForEachAddEnumerableMapping.cs [hunk 10]
```diff
@@ -14,24 +14,50 @@ namespace Riok.Mapperly.Descriptors.Mappings;
 public class ForEachAddEnumerableMapping : ExistingTargetMappingMethodWrapper
 {
     private readonly ObjectFactory? _objectFactory;
+    private readonly ITypeSymbol _typeToInstantiate;
+    private readonly string? _countPropertyName;
 
+    // warning: if countPropertyName is not null then typeToInstantiate must have a single integer constructor
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Mappings/ForEachAddEnumerableMapping.cs:20]: This comment applies to the second ctor but not to this one, does it? I'd also make it an xml comment to show it up in the IDE when invoking the constructor and reference the parameters with paramref.

**Review by latonz — CHANGES_REQUESTED** [review]
Thanks for the updates, one little nitpick.

**Comment by TimothyMakkison** [comment]
Thanks for the review 😄 

I noticed #621 could crop up in `TryBuildFastConversion`. I've updated the return types to use the correct element and added some unit tests

**Comment by latonz** [comment]
@TimothyMakkison sorry for merging the dependencies first (I thought there are open linter issues in this PR...), now this needs another rebase 🙄

**Review by latonz — APPROVED** [review]
